test/exchange_price_source.test.ts

```diff
@@ -0,0 +1,177 @@
+import { ExchangeAdapter, Ticker, Trade } from '../src/exchange_adapters/base'
+import BigNumber from 'bignumber.js'
+import { Exchange } from '../src/utils'
+import {
+  ExchangePriceSource,
+  OrientedAdapter,
+  PairData,
+  impliedPair,
+} from '../src/exchange_price_source'
+import { baseLogger } from '../src/default_config'
+import { MetricCollector } from '../src/metric_collector'
+
+jest.mock('../src/metric_collector')
+
+class MockAdapter implements ExchangeAdapter {
+  ticker: Ticker
+  pairSymbol: string
+  exchangeName: Exchange
+
+  constructor(ticker: Ticker) {
+    this.ticker = ticker
+    this.pairSymbol = ticker.symbol
+    this.exchangeName = ticker.source
+  }
+
+  async fetchTrades(): Promise<Trade[]> {
+    // Trades are not used.
+    return [] as Trade[]
+  }
+
+  async fetchTicker(): Promise<Ticker> {
+    return this.ticker
+  }
+}
+
+describe('impliedPair()', () => {
+  const testPair1: PairData = {
+    bid: new BigNumber(10.0),
+    ask: new BigNumber(10.01),
+    baseVolume: new BigNumber(10),
+    quoteVolume: new BigNumber(10),
+  }
+  const testPair2: PairData = {
+    bid: new BigNumber(2.0),
+    ask: new BigNumber(3.0),
+    baseVolume: new BigNumber(100),
+    quoteVolume: new BigNumber(100),
+  }
+  const testPair3: PairData = {
+    bid: new BigNumber(1.0),
+    ask: new BigNumber(1.0),
+    baseVolume: new BigNumber(100000),
+    quoteVolume: new BigNumber(100000),
+  }
+  const testCELOEUR: PairData = {
+    bid: new BigNumber(4.2),
+    ask: new BigNumber(4.21),
+    baseVolume: new BigNumber(5000),
+    quoteVolume: new BigNumber(21000),
+  }
+  const testEURUSD: PairData = {
+    bid: new BigNumber(1.21),
+    ask: new BigNumber(1.22),
+    baseVolume: new BigNumber(10000),
+    quoteVolume: new BigNumber(12100),
+  }
+
+  describe('single pair', () => {
+    it('calculates implied pair', () => {
+      const implied = impliedPair([testPair1])
+      expect(implied).toEqual(testPair1)
+    })
+  })
+
+  describe('two pairs', () => {
+    describe('test pairs', () => {
```

Comment:
any need for this describe?

Reply:
Just to separate the (simple) test pairs from the real world pairs.